snuba/datasets/transactions_processor.py

```diff
@@ -0,0 +1,115 @@
+from datetime import datetime
+
+import ipaddress
+import uuid
+
+from snuba.processor import (
+    _as_dict_safe,
+    MessageProcessor,
+    _ensure_valid_date,
+    _unicodify
+)
+from snuba.datasets.events_processor import (
+    enforce_retention,
+    extract_base,
+    extract_extra_contexts,
+    extract_extra_tags,
+    extract_user,
+)
+
+
+class TransactionsMessageProcessor(MessageProcessor):
+    PROMOTED_TAGS = [
+        "environment",
+        "sentry:release",
+        "sentry:user",
+        "sentry:dist",
+    ]
+
+    def __extract_timestamp(self, field):
+        timestamp = _ensure_valid_date(datetime.fromtimestamp(field))
+        milliseconds = int(timestamp.microsecond / 1000)
+        return (timestamp, milliseconds)
+
+    def process_message(self, message, metadata=None):
+        action_type = self.INSERT
+        processed = {'deleted': 0}
+        if not isinstance(message, (list, tuple)) and len(message) >= 2:
+            return None
+        version = message[0]
+        if version not in (0, 1, 2):
+            return None
+        type_, event = message[1:3]
+        if type_ != 'insert':
+            return None
+
+        data = event.get("data", {})
+        event_type = data.get("type")
+        if event_type != "transaction":
+            return None
+        base = extract_base(event)
+        enforce_retention(
+            event,
+            datetime.fromtimestamp(data['timestamp']),
+            base['retention_days'],
+            base['project_id'],
+        )
```

Comment:
It seems like there should be a better way to structure this that is less copy/paste here to extract the shared attributes between the events processor and this processor. I don't know as that should happen with this change, but seems like we should improve this later.

Reply:
refactored to avoid passing both project id and retention days to the method